File: assets/mcp/hw-notify.py
#!/usr/bin/env python3
"""
ClawEnv hw-notify — MCP Server for hardware device notifications (Python version)

Runs inside the sandbox (or native) as an MCP stdio server.
Agent calls notify() → this plugin POSTs to Bridge /api/hw/notify.

Usage:
  python3 hw-notify.py
  python3 hw-notify.py --bridge-url http://host.lima.internal:3100

Register with Hermes Agent:
  hermes mcp add hw-notify --config '{"command":"python3","args":["/workspace/hw-notify/notify.py"]}'
"""
from __future__ import annotations

import sys
import os
import json
import argparse
import asyncio
from typing import Optional

import httpx
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import TextContent

parser = argparse.ArgumentParser()
parser.add_argument("--bridge-url", default="")
args = parser.parse_args()

bridge_url = args.bridge_url

# Auto-detect bridge URL (same logic as mcp-bridge)
if not bridge_url:
    host_ip = os.environ.get("CLAWENV_HOST_IP")
    if host_ip:
        bridge_url = f"http://{host_ip}:3100"
    else:
        try:
            r = httpx.get("http://host.containers.internal:3100/api/health", timeout=2)
            if r.status_code == 200:
                bridge_url = "http://host.containers.internal:3100"
        except Exception:
            pass
        if not bridge_url:
            bridge_url = "http://host.lima.internal:3100"

print(f"[hw-notify] Bridge URL: {bridge_url}", file=sys.stderr)

client = httpx.AsyncClient(base_url=bridge_url, timeout=10)

server = Server("hw-notify")


def text(s: str) -> list:
    return [TextContent(type="text", text=s)]


@server.tool()
async def notify(
    message: str,
    level: Optional[str] = "info",
    device_id: Optional[str] = "*",
) -> list:
    """Send a notification to connected hardware devices. Use this when you need to alert the user on their hardware device (e.g., task completed, action required, important update)."""
    resp = await client.post("/api/hw/notify", json={
        "message": message,
        "level": level or "info",
        "device_id": device_id or "*",
        "from_instance": os.environ.get("CLAWENV_INSTANCE", ""),
    })
    if resp.status_code != 200:
        return text(f"Notification failed ({resp.status_code}): {resp.text}")
    return text(f"Notification sent: {message}")


async def main():
    async with stdio_server() as (read_stream, write_stream):
        print("[hw-notify] Server started on stdio", file=sys.stderr)
        await server.run(read_stream, write_stream, server.create_initialization_options())


if __name__ == "__main__":
    asyncio.run(main())

File: v2/assets/mcp/hw-notify.mjs
#!/usr/bin/env node
/**
 * ClawEnv hw-notify — MCP Server for hardware device notifications
 *
 * Runs inside the sandbox (or native) as an MCP stdio server.
 * Agent calls notify() → this plugin POSTs to Bridge /api/hw/notify.
 *
 * Usage:
 *   node hw-notify.mjs
 *   node hw-notify.mjs --bridge-url http://host.lima.internal:3100
 *
 * Register with OpenClaw:
 *   openclaw mcp set hw-notify '{"command":"node","args":["/workspace/hw-notify/index.mjs"]}'
 */
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

// Parse CLI args
const args = process.argv.slice(2);
let bridgeUrl = "";
for (let i = 0; i < args.length; i++) {
    if (args[i] === "--bridge-url" && args[i + 1]) {
        bridgeUrl = args[i + 1];
        i++;
    }
}

// Auto-detect bridge URL (same logic as mcp-bridge)
if (!bridgeUrl) {
    const hostIp = process.env.CLAWENV_HOST_IP;
    if (hostIp) {
        bridgeUrl = `http://${hostIp}:3100`;
    } else {
        try {
            const resp = await fetch("http://host.containers.internal:3100/api/health", { signal: AbortSignal.timeout(2000) });
            if (resp.ok) bridgeUrl = "http://host.containers.internal:3100";
        } catch { /* not Podman */ }
        if (!bridgeUrl) {
            bridgeUrl = "http://host.lima.internal:3100";
        }
    }
}
console.error(`[hw-notify] Bridge URL: ${bridgeUrl}`);

const server = new McpServer({ name: "hw-notify", version: "0.1.0" });

server.tool(
    "notify",
    "Send a notification to connected hardware devices. Use this when you need to alert the user on their hardware device (e.g., task completed, action required, important update).",
    {
        message: z.string().describe("Notification message content"),
        level: z.enum(["info", "alert", "action"]).optional().describe("Urgency level: info (default), alert, or action (requires user response)"),
        device_id: z.string().optional().describe("Target device ID, omit or '*' to broadcast to all devices"),
    },
    async ({ message, level, device_id }) => {
        const resp = await fetch(`${bridgeUrl}/api/hw/notify`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                message,
                level: level || "info",
                device_id: device_id || "*",
                from_instance: process.env.CLAWENV_INSTANCE || "",
            }),
        });
        if (!resp.ok) {
            const text = await resp.text();
            return { content: [{ type: "text", text: `Notification failed (${resp.status}): ${text}` }] };
        }
        return { content: [{ type: "text", text: `Notification sent: ${message}` }] };
    }
);

const transport = new StdioServerTransport();
await server.connect(transport);
console.error("[hw-notify] Server started on stdio");
